nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/RunMongoAggregation.java

```diff
@@ -0,0 +1,210 @@
+        List<Map> values = mapper.readValue(query, List.class);
+        for (Map val : values) {
+            result.add(new BasicDBObject(val));
+        }
+
+        return result;
+    }
+
+    public static final Validator AGG_VALIDATOR = (subject, value, context) -> {
+        final ValidationResult.Builder builder = new ValidationResult.Builder();
+        builder.subject(subject).input(value);
+
+        if (context.isExpressionLanguageSupported(subject) && context.isExpressionLanguagePresent(value)) {
+            return builder.valid(true).explanation("Contains Expression Language").build();
+        }
+
+        String reason = null;
+        try {
+            buildAggregationQuery(value);
+        } catch (final RuntimeException | IOException e) {
+            reason = e.getLocalizedMessage();
+        }
+
+        return builder.explanation(reason).valid(reason == null).build();
+    };
+
+    static final PropertyDescriptor QUERY = new PropertyDescriptor.Builder()
+            .name("mongo-agg-query")
+            .displayName("Query")
+            .expressionLanguageSupported(true)
+            .description("The aggregation query to be executed.")
+            .required(true)
+            .addValidator(AGG_VALIDATOR)
+            .build();
+
+    static {
+        List<PropertyDescriptor> _propertyDescriptors = new ArrayList<>();
+        _propertyDescriptors.addAll(descriptors);
+        _propertyDescriptors.add(CHARSET);
+        _propertyDescriptors.add(QUERY);
+        _propertyDescriptors.add(QUERY_ATTRIBUTE);
+        _propertyDescriptors.add(BATCH_SIZE);
+        _propertyDescriptors.add(RESULTS_PER_FLOWFILE);
+        _propertyDescriptors.add(SSL_CONTEXT_SERVICE);
+        _propertyDescriptors.add(CLIENT_AUTH);
+        propertyDescriptors = Collections.unmodifiableList(_propertyDescriptors);
+
+        final Set<Relationship> _relationships = new HashSet<>();
+        _relationships.add(REL_RESULTS);
+        _relationships.add(REL_ORIGINAL);
+        _relationships.add(REL_FAILURE);
+        relationships = Collections.unmodifiableSet(_relationships);
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return propertyDescriptors;
+    }
+
+    static String buildBatch(List batch) {
+        ObjectMapper mapper = new ObjectMapper();
+        String retVal;
+        try {
+            retVal = mapper.writeValueAsString(batch.size() > 1 ? batch : batch.get(0));
+        } catch (Exception e) {
+            retVal = null;
+        }
+
+        return retVal;
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        FlowFile flowFile = null;
+        if (context.hasIncomingConnection()) {
+            flowFile = session.get();
+
+            if (flowFile == null) {
```

Comment:
You're not checking for context.hasNonLoopConnection() here, is that on purpose? If I have no upstream connections but I connect Failure back to the processor, then it won't run my aggregation. If this is on purpose, please mention it in the relationship description(s) and/or the main capability description. Otherwise, you can add hasNonLoopConnection() (see ExecuteSQL for an example with comments). I prefer the latter, but it is up to you.

Reply:
Oversight on my part. Made the change.